File: Components/search_cma.m
function [output1,output2] = search_cma(varargin)
% The evolution strategy with convariance matrix adaption.

%------------------------------Reference-----------------------------------
% N. Hansen and A. Ostermeier, Completely derandomized selfadaptation in
% evolution strategies, Evolutionary Computation, 2001, 9(2): 159-195.
%------------------------------Copyright-----------------------------------
% Copyright (C) <2023>  <Swarm Intelligence Lab>

% AutoOptLib is a free software. You can use, redistribute, and/or modify
% it under the terms of the GNU General Public License as published by the 
% Free Software Foundation, either version 3 of the License, or any later 
% version. 

% Please reference the paper below if using AutoOptLib in your publication:
% @article{zhao2023autooptlib,
%  title={AutoOptLib: A Library of Automatically Designing Metaheuristic 
%         Optimization Algorithms in Matlab},
%  author={Zhao, Qi and Yan, Bai and Hu, Taiwei and Chen, Xianglong and 
%          Yang, Jian and Shi, Yuhui},
%  journal={arXiv preprint 	arXiv:2303.06536},
%  year={2023}
% }
%--------------------------------------------------------------------------

mode = varargin{end};
switch mode
    case 'execute'
        Parent  = varargin{1};
        Problem = varargin{2};
        lower   = Problem.bound(1,:);
        upper   = Problem.bound(2,:);
        Aux     = varargin{4};
        if ~isnumeric(Parent)
            Parent  = Parent.decs;
        end
        [N,D] = size(Parent);

        % initialize parameters
        if ~isfield(Aux,'cma_halfN')
            Aux.cma_halfN   = round(N/2);
            w               = log(Aux.cma_halfN+0.5)-log(1:Aux.cma_halfN);
            Aux.cma_w       = w./sum(w);
            Aux.cma_betterN = 1/sum(Aux.cma_w.^2);

            Aux.cma_csigma  = (Aux.cma_betterN+2)/(D+Aux.cma_betterN+5);
            Aux.cma_dsigma  = Aux.cma_csigma+2*max(sqrt((Aux.cma_betterN-1)/(D+1))-1,0)+1;
            Aux.cma_chiN     = sqrt(D)*(1-1/(4*D)+1/(21*D^2));

            Aux.cma_cc    = (4+Aux.cma_betterN/D)/(4+D+2*Aux.cma_betterN/D);
            Aux.cma_ccov  = 2/((D+1.3)^2+Aux.cma_betterN);
            Aux.cma_cmu   = min(1-Aux.cma_ccov,2*(Aux.cma_betterN-2+1/Aux.cma_betterN)/((D+2)^2+2*Aux.cma_betterN/2));
            Aux.cma_hth   = (1.4+2/(D+1))*Aux.cma_chiN;

            Aux.cma_mean  = unifrnd(lower,upper);
            Aux.cma_ps    = zeros(1,D);
            Aux.cma_pc    = zeros(1,D);
            Aux.cma_C     = eye(D);
            Aux.cma_sigma = 0.1*(upper-lower);
        end

        % load parameters
        C     = Aux.cma_C;
        mean  = Aux.cma_mean;
        sigma = Aux.cma_sigma;

        % search
        Disturb = zeros(N,D);
        for i = 1:N
            Disturb(i,:) = mvnrnd(zeros(1,D),C);
        end
        output1 = mean+sigma.*Disturb;
        Aux.cma_Disturb = Disturb;
        output2 = Aux;

    case 'parameter'
        % n/a

    case 'behavior'
        output1 = {'';'GS'}; % always performs global search

    case 'algorithm'
        Parent = varargin{1};
        bound  = varargin{2};
        Aux    = varargin{3};
        lower  = bound(1,:);
        upper  = bound(2,:);
        [N,D] = size(Parent);

        % initialize parameters
        if ~isfield(Aux,'cma_halfN')
            Aux.cma_halfN   = round(N/2);
            w               = log(Aux.cma_halfN+0.5)-log(1:Aux.cma_halfN);
            Aux.cma_w       = w./sum(w);
            Aux.cma_betterN = 1/sum(Aux.cma_w.^2);

            Aux.cma_csigma  = (Aux.cma_betterN+2)/(D+Aux.cma_betterN+5);
            Aux.cma_dsigma  = Aux.cma_csigma+2*max(sqrt((Aux.cma_betterN-1)/(D+1))-1,0)+1;
            Aux.cma_chiN     = sqrt(D)*(1-1/(4*D)+1/(21*D^2));

            Aux.cma_cc    = (4+Aux.cma_betterN/D)/(4+D+2*Aux.cma_betterN/D);
            Aux.cma_ccov  = 2/((D+1.3)^2+Aux.cma_betterN);
            Aux.cma_cmu   = min(1-Aux.cma_ccov,2*(Aux.cma_betterN-2+1/Aux.cma_betterN)/((D+2)^2+2*Aux.cma_betterN/2));
            Aux.cma_hth   = (1.4+2/(D+1))*Aux.cma_chiN;

            Aux.cma_mean  = unifrnd(lower,upper);
            Aux.cma_ps    = zeros(1,D);
            Aux.cma_pc    = zeros(1,D);
            Aux.cma_C     = eye(D);
            Aux.cma_sigma = 0.1*(upper-lower);
        end

        % load parameters
        C     = Aux.cma_C;
        mean  = Aux.cma_mean;
        sigma = Aux.cma_sigma;

        % search
        Disturb = zeros(N,D);
        for i = 1:N
            Disturb(i,:) = mvnrnd(zeros(1,D),C);
        end
        output1 = mean+sigma.*Disturb;
        Aux.cma_Disturb = Disturb;
        output2 = Aux;
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: src/autooptlib/components/search_cma.py
"""Python translation of search_cma."""
from __future__ import annotations

from typing import Any, Tuple

import numpy as np

from ._utils import flex_get


def _extract_decs(parent_obj: Any) -> np.ndarray:
    if isinstance(parent_obj, np.ndarray):
        return parent_obj
    decs = flex_get(parent_obj, "decs")
    if callable(decs):
        return np.asarray(decs(), dtype=float)
    if decs is not None:
        return np.asarray(decs, dtype=float)
    # assume sequence of Solution objects
    return np.vstack([np.asarray(flex_get(item, "dec"), dtype=float) for item in parent_obj])


def _ensure_aux(aux: Any) -> dict:
    if aux is None or not isinstance(aux, dict):
        return {}
    return aux


def _init_cma(aux: dict, n: int, d: int, lower: np.ndarray, upper: np.ndarray) -> None:
    half_n = int(np.round(n / 2))
    w = np.log(np.arange(1, half_n + 1) + 0.5)
    w = w - w[-1]
    w = w[::-1]
    w = w / np.sum(w)
    better_n = 1.0 / np.sum(w ** 2)

    aux.setdefault("cma_halfN", half_n)
    aux.setdefault("cma_w", w)
    aux.setdefault("cma_betterN", better_n)
    aux.setdefault("cma_csigma", (better_n + 2) / (d + better_n + 5))
    aux.setdefault(
        "cma_dsigma",
        aux["cma_csigma"] + 2 * max(np.sqrt((better_n - 1) / (d + 1)) - 1, 0) + 1,
    )
    aux.setdefault("cma_chiN", np.sqrt(d) * (1 - 1 / (4 * d) + 1 / (21 * d * d)))
    aux.setdefault("cma_cc", (4 + better_n / d) / (4 + d + 2 * better_n / d))
    aux.setdefault("cma_ccov", 2 / ((d + 1.3) ** 2 + better_n))
    aux.setdefault(
        "cma_cmu",
        min(
            1 - aux["cma_ccov"],
            2 * (better_n - 2 + 1 / better_n) / ((d + 2) ** 2 + 2 * better_n / 2),
        ),
    )
    aux.setdefault("cma_hth", (1.4 + 2 / (d + 1)) * aux["cma_chiN"])
    aux.setdefault("cma_mean", lower + (upper - lower) * np.random.rand(d))
    aux.setdefault("cma_ps", np.zeros(d))
    aux.setdefault("cma_pc", np.zeros(d))
    aux.setdefault("cma_C", np.eye(d))
    aux.setdefault("cma_sigma", 0.1 * (upper - lower))


def _sample(aux: dict, n: int, d: int) -> Tuple[np.ndarray, np.ndarray]:
    mean = np.asarray(aux["cma_mean"], dtype=float)
    sigma = np.asarray(aux["cma_sigma"], dtype=float)
    C = np.asarray(aux["cma_C"], dtype=float)
    disturbance = np.random.multivariate_normal(np.zeros(d), C, size=n)
    offspring = mean + sigma * disturbance
    aux["cma_Disturb"] = disturbance
    return offspring, aux


def search_cma(*args):
    mode = args[-1]
    if mode == "execute":
        parent_obj = args[0]
        problem = args[1]
        aux = _ensure_aux(args[3] if len(args) > 3 else None)
        parent = _extract_decs(parent_obj)
        n, d = parent.shape
        bound = flex_get(problem, "bound")
        lower = np.asarray(bound[0], dtype=float)
        upper = np.asarray(bound[1], dtype=float)
        _init_cma(aux, n, d, lower, upper)
        offspring, aux = _sample(aux, n, d)
        return offspring, aux

    if mode == "parameter":
        return None, None

    if mode == "behavior":
        return ["", "GS"], None

    if mode == "algorithm":
        parent = np.asarray(args[0], dtype=float)
        bound = np.asarray(args[1], dtype=float)
        aux = _ensure_aux(args[2] if len(args) > 2 else None)
        n, d = parent.shape
        lower = bound[0]
        upper = bound[1]
        _init_cma(aux, n, d, lower, upper)
        offspring, aux = _sample(aux, n, d)
        return offspring, aux

    raise ValueError(f"Unsupported mode: {mode}")